src/accountant.rs

```diff
@@ -131,23 +133,34 @@ impl Accountant {
         // Hold a write lock before the condition check, so that a debit can't occur
         // between checking the balance and the withdraw.
         let option = bals.get(&tr.from);
+
         if option.is_none() {
             return Err(AccountingError::AccountNotFound);
         }
-        let mut bal = option.unwrap().write().unwrap();
 
         if !self.reserve_signature_with_last_id(&tr.sig, &tr.data.last_id) {
             return Err(AccountingError::InvalidTransferSignature);
         }
 
-        if *bal < tr.data.tokens {
+        let bal = option.unwrap();
+        let current = bal.load(Ordering::Relaxed) as i64;
+
+        if current < tr.data.tokens {
```

Comment:
I was thinking you could start a loop here that continues until `compare_exchange` returns `Ok` or the balance check returns `InsufficentFunds`. You shouldn't need to add that new error type `BalanceUpdatedBeforeTransactionCompleted`.